Read depletion_chain information from main cross section file

drewejohnson opened the pull request:
This PR closes #1239 by replacing the environment variable `OPENMC_DEPLETE_CHAIN` in favor of a `depletion_chain` node in the main `OPENMC_CROSS_SECTION` file. Unless a `chain_file` argument is directly passed when constructing an `openmc.deplete.Operator` object, the chain is created using this new default file. A `DeprecationWarning` is raised alerting the user to the change in behavior.

A unit test was added that makes a dummy `cross_sections.xml` file with this node, swaps the `OPENMC_CROSS_SECTION` environment variable to use this file, and creates a `Operator`. The depletion chain on this operator is that from `chain_simple.xml`, and is compared to a `Chain` created directly from this file.

paulromano commented on tests/unit_tests/test_deplete_operator.py:
```diff
@@ -0,0 +1,72 @@
+        out.write(bare_xs_contents)
+
+    yield
+    remove(BARE_XS_FILE)
```
There are two options in place for using a temporary directory that you can use instead of manually removing this:
1. In `tests/conftest.py`, there is a fixture called `run_in_tmpdir` that will run a test in a temporary directory and then switch back to wherever the cwd was before.
2. In `tests/__init__.py`, there is a `cdtemp` function that can be using as a context manager to temporary change into a temporary directory.

drewejohnson replied:
Will use the run_in_tmpdir option

paulromano commented on tests/unit_tests/test_deplete_operator.py:
```diff
@@ -0,0 +1,72 @@
+from os import environ
+from unittest import mock
+from pathlib import Path
+import pytest
```
Move the `pytest` import together with the `openmc` imports

paulromano commented on tests/unit_tests/test_deplete_operator.py:
```diff
@@ -0,0 +1,72 @@
+from openmc.deplete.chain import Chain
+
+BARE_XS_FILE = "bare_cross_sections.xml"
+CHAIN_PATH = Path().cwd() / "tests" / "chain_simple.xml"
```
This only works if the current working directory is correct. I believe you could use `Path(__file__).parents[1] / "chain_simple.xml"` for it to work anywhere.

drewejohnson commented on openmc/deplete/abc.py:
```diff
@@ -62,8 +65,23 @@ def __init__(self, chain_file=None):
+                    raise IOError(
+                        "OPENMC_CROSS_SECTIONS environment variable "
+                        "not set.")
+                root = ET.parse(xs_file).getroot()
```
DataLibrary can now handle a `cross_sections` xml file with a `depletion_chain` node, for registering, reading, and exporting. The library dictionary has an empty list for the `materials` attribute.

drewejohnson commented:
Thanks for the feedback @paulromano! I was out on vacation for a week, but I believe I've addressed your comments. The `DataLibrary` changes are what I'm most unsure about, so I tried to find the most consistent way to include information about the depletion chain

drewejohnson commented:
Looks like all but one test passed, with python 3.6 `omp=mpi=phdf5=n` timed out. https://travis-ci.org/openmc-dev/openmc/jobs/544354481#L403. 

[Apparently the njoy install hung.](https://travis-ci.org/openmc-dev/openmc/jobs/544354481#L396-L401)

paulromano commented on openmc/deplete/abc.py:
```diff
@@ -62,8 +66,21 @@ def __init__(self, chain_file=None):
-                              "environment variable OPENMC_DEPLETE_CHAIN.")
+                data = DataLibrary.from_xml()
+                # search for depletion_chain path from end of list
+                for lib in data.libraries[::-1]:
```
`for lib in reversed(data.libraries):`

paulromano commented on openmc/data/library.py:
```diff
@@ -48,19 +49,30 @@ def register_file(self, filename):
+                .format(path.name, self.__class__.__name__))
 
-        library = {'path': filename, 'type': filetype, 'materials': materials}
+        library = {'path': path.name, 'type': filetype, 'materials': materials}
```
Using `path.name` looks like it loses information. If I take an existing `cross_sections.xml` that doesn't have a depletion chain in it, load it with `DataLibrary.from_xml`, register a chain file, and then re-export it, the exported XML file doesn't include the full path. You can probably use `str(path)` here instead.

Alternatively (this would require more work), the dictionary can just store a `Path` object under the 'path' key instead of a string.

drewejohnson replied:
Interesting. I'll add some unit tests to cover this case in the next push

paulromano commented:
There's one more instance of `OPENMC_DEPLETE_CHAIN` in `openmc/model/model.py` to change.

paulromano commented:
Also forgot to mention before -- I restarted the Travis job and everything was good the second time around :+1:

paulromano commented on openmc/model/model.py:
```diff
@@ -133,8 +133,9 @@ def deplete(self, timesteps, chain_file=None, method='cecm',
-            :envvar:`OPENMC_DEPLETE_CHAIN` environment variable if it exists.
+            Path to the depletion chain XML file.  Defaults to the chain
+            found under the ``depletion_chain`` in the
+            :envvar:`OPENMC_CROSS_SECITON` environment variable if it exists.
```
OPENMC_CROSS_SECTIONS

paulromano commented:
Good to go after that one last comment is resolved!

paulromano commented:
Congrats on your first PR @drewejohnson! Oh, and see you next week :wave:
